guardian/cdk
# Add pattern for scheduled lambdas

## What does this change?

This PR is a first attempt at a scheduled lambda pattern. It also introduces a construct for CloudWatch alarms, and a pattern for a lambda-specific CloudWatch alarms, based on error %. 

I've introduced the CloudWatch-y stuff at the same time as the scheduled lambda pattern because a number of examples of this pattern 'in the wild' also include an alert. I think we should be encouraging developers to think about monitoring their software whenever they add new compute resources, so hopefully this will help to do that.

## Does this change require changes to existing projects or CDK CLI?

No changes are required, however stacks that use schedule-based lambdas can simplify their infrastructure definition slightly. 

## How to test

1. Upgrade an existing stack to use this new pattern. The snapshot tests should show 0 (or trivial) changes.
2. Create a new stack based on this pattern. You should end up with a functional scheduled lambda.
3. Configure monitoring and force the lambda to fail X% of the time to ensure that the new alert pattern works as expected.

I've tested 2 & 3 in the `developerPlayground` account.

## How can we measure success?

* Projects already using this library (e.g. `tag-janitor` and `pagerduty-summary`) should be able to simplify their infrastructure definition slightly
* New projects should get sensible monitoring for a tiny amount of extra code/effort

## Have we considered potential risks?

* This is all new functionality; merging this PR should not have an impact on any projects currently using this library.
